### DISCUSSION TITLE
修改编译链默认的cxx

### DISCUSSION THREAD

**Discussion by zt449569708** [xmake-io/xmake#2254] [Q&A]
### Xmake Version

v2.6.4+master.26e6cc713

### Operating System Version and Architecture

ubuntu18.04

### Describe Bug

```
toolchain("clang")
    set_kind("standalone")
    set_sdkdir("/usr/local/xxx")
    on_load(function (toolchain)
        toolchain:load_cross_toolchain()
        toolchain:set("toolset", "cxx", "clang++")
    end)
toolchain_end()
```

自定义了一个clang编译工具，但是想修改默认的cxx为clang++，但是实际运行的时候，改不成功，如图：
![image](https://user-images.githubusercontent.com/6011216/162354780-6ee8709b-537b-4c7c-94da-d9b15631976c.png)

### Expected Behavior

cxx成功修改为clang++

**Comment by waruqi** [comment]
clang 是内置工具链，换个名字， 另外直接 xmake f --cxx=clang++ 不就完事了么

**Comment by zt449569708** [comment]
clang不是内置工具链，如果xmake f --cxx=clang++ 这样用的话，ld的时候用的就是g++了，因为gcc才是系统默认的内置工具链。


 
---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2022年4月8日 12:14:55
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [xmake-io/xmake] 修改编译链默认的cxx (Issue #2250)




 
clang 是内置工具链，换个名字， 另外直接 xmake f --cxx=clang++ 不就完事了么

**Comment by waruqi** [comment]
我是说，toolchain("clang") 内部已经定义掉了。。你外面定义有冲突，改个名

https://github.com/xmake-io/xmake/blob/master/xmake/toolchains/clang/xmake.lua

**Comment by zt449569708** [comment]
```
toolchain("my-clang")
    set_kind("standalone")
    set_sdkdir("/usr/local/clang")
toolchain_end()

if is_plat("myplat") then
    set_toolchains("my-clang")
end
```

改了名字，编译的时候，找不到clang,。步骤先是运行`xmake f -p myplat -c`，然后`xmake -rv`输出如下：
![image](https://user-images.githubusercontent.com/6011216/162445885-250638f0-64ea-4860-bcd4-96ed16a83dc0.png)

**Comment by waruqi** [comment]
>     set_sdkdir("/usr/local/clang")

这个是目录？clang在 `/usr/local/clang/bin/clang`? 

而且你的 toolchain 里面什么都没设置，cxx 也没设置，默认找的当然还是 gcc 了

**Comment by waruqi** [comment]
而且我说了，你要改 cxx 。。直接 `xmake f --cxx=clang++`  就完事了。或者直接 `set_toolset("cxx", "clang++")` ，toolchain 都没必要去定义。

**Comment by L-Sun** [comment]
不太清楚为什么clang的优先级比clang++更高
https://github.com/xmake-io/xmake/blob/df7c435cc7e09a2aa834289f534e04195872140f/xmake/toolchains/clang/xmake.lua#L35
